### ISSUE TITLE
36. **Gas optimisations from constant handling of keccak expressions**

### ISSUE THREAD

**Issue by RusseII** [porter-finance/v1-core#221]
36. **Gas optimisations from constant handling of keccak expressions**
    141. **Summary**: keccak expression assigned to a constant variable is recomputed each time.
    142. **Details**: Due to how constant variables are implemented (replacements at compile-time), a keccak expression assigned to a constant variable is recomputed each time that the variable is used, which is gas inefficient.
    143. **Github Permalinks**: [WITHDRAW_ROLE](https://github.com/porter-finance/v1-core/blob/CAREX/contracts/Bond.sol#L75), [MINT_ROLE](https://github.com/porter-finance/v1-core/blob/CAREX/contracts/Bond.sol#L82), [ISSUER_ROLE](https://github.com/porter-finance/v1-core/blob/CAREX/contracts/BondFactory.sol#L18)
    144. **Mitigation: **Replace the “keccak” expressions by their hardcoded value in the constant declarations and document in a comment about how their value was calculated.

**Comment by Namaskar-1F64F** [comment]
using the following code and the gas costs - it seems like this is incorrect.

```
// SPDX-License-Identifier: GPL-3.0

pragma solidity 0.8.9;

contract A {
    // bytes32 public constant WITHDRAW_ROLE = keccak256("WITHDRAW_ROLE"); 
    // bytes32 public constant WITHDRAW_ROLE = 0x5d8e12c39142ff96d79d04d15d1ba1269e4fe57bb9d26f43523628b34ba108ec;
    // bytes32 public immutable WITHDRAW_ROLE = 0x5d8e12c39142ff96d79d04d15d1ba1269e4fe57bb9d26f43523628b34ba108ec;
    // bytes32 public immutable WITHDRAW_ROLE = keccak256("WITHDRAW_ROLE"); 

    // deploy
    // 115009 
    // 116287
    // 115217
    // 115175

    // run
    //21379 
    //21385
    //21379
    //21379 
    function retrieve() public returns (bytes32) {
        return WITHDRAW_ROLE;
    }
}
```

**Comment by RusseII** [comment]
https://canary.discord.com/channels/814328279468474419/954594497378025563/963187773458747402

Looks like was fixed in [0.6.12](https://github.com/ethereum/solidity/blob/develop/Changelog.md#0612-2020-07-22)! Nice catch

<img width="536" alt="image" src="https://user-images.githubusercontent.com/15036618/162975455-cb3a2a31-1d01-42f0-beaf-d7c1986ed87f.png">
